File: bat_grade.py
"""
Bat_Grade — Nuke-style colour grading with a live canvas preview.

Mirrors Nuke's Grade node math:
    base    = (in - blackpoint) / max(whitepoint - blackpoint, 1e-6)
    leveled = base * (gain - lift) + lift
    out     = (leveled * multiply + offset) ** (1 / max(gamma, 1e-6))
              (negative values skip the gamma and pass through linear)
    out     = clamp(out, max=1) if clamp_white; clamp(out, min=0) if clamp_black

An optional MASK input gates the grade: where mask == 1 the grade
applies fully, where mask == 0 the pixels pass through unchanged
(linear mix between them).

The live preview on the node body is driven by the sibling
``web/bat_grade.js`` — Python pushes a downscaled thumbnail of the
input image (and, if connected, the mask) back through the
``{"ui": ...}`` channel on every run, so the JS canvas can apply
the grade formula in real time as the artist drags the sliders.
The actual graded output is computed here on the next run.
"""

import base64
from io import BytesIO

import numpy as np
import torch
from PIL import Image

from .bat_crop import _broadcast_mask_to_n
from .bat_hdr_preview import hdr_tile


def _first(t: torch.Tensor) -> torch.Tensor:
    return t[0:1] if t is not None and t.shape[0] > 1 else t


def _b64_jpeg(arr_hwc: np.ndarray, max_dim: int = 384, quality: int = 80) -> str:
    """Downscaled base64 JPEG of an (H,W,3) uint8 numpy array."""
    im = Image.fromarray(arr_hwc, "RGB")
    if max(im.size) > max_dim:
        r = max_dim / max(im.size)
        im = im.resize((max(1, int(im.width * r)), max(1, int(im.height * r))),
                       Image.BILINEAR)
    buf = BytesIO()
    im.save(buf, format="JPEG", quality=quality)
    return base64.b64encode(buf.getvalue()).decode("utf-8")


def _b64_png_l(arr_hw: np.ndarray, max_dim: int = 384) -> str:
    """Downscaled base64 PNG of an (H,W) uint8 mask array."""
    im = Image.fromarray(arr_hw, "L")
    if max(im.size) > max_dim:
        r = max_dim / max(im.size)
        im = im.resize((max(1, int(im.width * r)), max(1, int(im.height * r))),
                       Image.BILINEAR)
    buf = BytesIO()
    im.save(buf, format="PNG", optimize=False, compress_level=3)
    return base64.b64encode(buf.getvalue()).decode("utf-8")


def _apply_grade(
    image: torch.Tensor,      # (N, H, W, 3) float in [0,1]
    mask: torch.Tensor,       # (N, H, W)    float in [0,1], or None
    blackpoint: float,
    whitepoint: float,
    lift: float,
    gain: float,
    multiply: float,
    offset: float,
    gamma: float,
    clamp_white: bool,
    clamp_black: bool,
) -> torch.Tensor:
    """Apply Nuke's Grade formula. See module docstring for the math."""
    img = image.to(torch.float32)
    wp_minus_bp = max(whitepoint - blackpoint, 1e-6)
    g = max(gamma, 1e-6)

    # Pass-through params (the node defaults) are an exact identity; skip the
    # full-batch arithmetic entirely.
    if (blackpoint == 0.0 and whitepoint == 1.0 and lift == 0.0 and gain == 1.0
            and multiply == 1.0 and offset == 0.0 and g == 1.0
            and not clamp_white and not clamp_black):
        return img

    # The levels stage is affine in `in`, so fold it into ONE multiply-add:
    #   ((in - bp)/(wp - bp) * (gain - lift) + lift) * multiply + offset
    #   = in * A + B
    # and do everything after it in place. The step-by-step version kept
    # base / leveled / out alive together: ~6x the batch at peak versus ~2x.
    a = (gain - lift) / wp_minus_bp * multiply
    b = (lift - blackpoint * (gain - lift) / wp_minus_bp) * multiply + offset
    out = img * a
    out.add_(b)

    # pow on negatives is undefined. Negatives pass through LINEAR (Nuke's
    # behaviour) rather than being zeroed, so clamp_black is a real toggle:
    # it used to be a no-op because this step already clamped them to 0.
    if g != 1.0:
        if clamp_black:
            out.clamp_(min=0.0).pow_(1.0 / g)
        else:
            neg = out < 0
            neg_vals = out[neg]
            out.clamp_(min=0.0).pow_(1.0 / g)
            out[neg] = neg_vals

    if clamp_white:
        out.clamp_(max=1.0)
    if clamp_black:
        out.clamp_(min=0.0)

    if mask is not None:
        # A mask batch that is neither 1 nor N used to crash the mix below
        # ("size of tensor a (10) must match … (5)"); hold its last frame /
        # truncate, the same way Bat_Crop brings a mask to the image batch.
        n = img.shape[0]
        if mask.shape[0] not in (1, n):
            mask = _broadcast_mask_to_n(mask, n)
        # mask: (N, H, W) → (N, H, W, 1) for broadcasting against (N,H,W,3)
        m = mask.to(torch.float32).unsqueeze(-1).clamp(0, 1)
        # Resize mask to image spatial size if they differ (e.g. mask is
        # 1/2 res after a downscale).
        if m.shape[1:3] != img.shape[1:3]:
            # NCHW for interpolate
            m_nchw = m.permute(0, 3, 1, 2)
            m_nchw = torch.nn.functional.interpolate(
                m_nchw, size=img.shape[1:3], mode="bilinear", align_corners=False,
            )
            m = m_nchw.permute(0, 2, 3, 1)
        # out*m + img*(1-m), in place.
        out.sub_(img).mul_(m).add_(img)
    return out


class BatGrade:
    """Nuke-style grade with a live canvas preview."""

    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "image":       ("IMAGE",),
                "blackpoint":  ("FLOAT", {"default": 0.0, "min": -1.0, "max": 1.0, "step": 0.001}),
                "whitepoint":  ("FLOAT", {"default": 1.0, "min": 0.001, "max": 4.0, "step": 0.001}),
                "lift":        ("FLOAT", {"default": 0.0, "min": -1.0, "max": 1.0, "step": 0.001}),
                "gain":        ("FLOAT", {"default": 1.0, "min": -4.0, "max": 4.0, "step": 0.001}),
                "multiply":    ("FLOAT", {"default": 1.0, "min": 0.0,  "max": 4.0, "step": 0.001}),
                "offset":      ("FLOAT", {"default": 0.0, "min": -1.0, "max": 1.0, "step": 0.001}),
                "gamma":       ("FLOAT", {"default": 1.0, "min": 0.01, "max": 4.0, "step": 0.001}),
                "clamp_white": ("BOOLEAN", {"default": False}),
                "clamp_black": ("BOOLEAN", {"default": False}),
                # Which frame of the input batch the live preview uses.
                # Doesn't affect the actual output (the grade still
                # applies to every frame). Lets the artist pick a more
                # representative frame than the default frame 0 — e.g.
                # the middle of a shot where the action is.
                "preview_frame": ("INT", {"default": 0, "min": 0, "max": 9999}),
            },
            "optional": {
                "mask": ("MASK",),
            },
        }

    RETURN_TYPES = ("IMAGE",)
    RETURN_NAMES = ("image",)
    FUNCTION = "grade"
    CATEGORY = "BAT/Colour"
    DESCRIPTION = (
        "Nuke-style grade with blackpoint/whitepoint, lift/gain, "
        "multiply, offset, and gamma. Optional mask gates the grade. "
        "Live canvas preview on the node body."
    )

    def grade(self, image, blackpoint, whitepoint, lift, gain, multiply,
              offset, gamma, clamp_white, clamp_black, preview_frame=0, mask=None):
        out = _apply_grade(
            image=image,
            mask=mask,
            blackpoint=float(blackpoint),
            whitepoint=float(whitepoint),
            lift=float(lift),
            gain=float(gain),
            multiply=float(multiply),
            offset=float(offset),
            gamma=float(gamma),
            clamp_white=bool(clamp_white),
            clamp_black=bool(clamp_black),
        )

        # Push a preview thumbnail of the SELECTED frame to the JS
        # canvas so it can apply the grade live as sliders move. The
        # frame index is clamped to the batch length so an aspirational
        # widget value (say "frame 200" on a 24-frame clip) falls back
        # to the last frame instead of erroring.
        idx = max(0, min(int(preview_frame), image.shape[0] - 1))
        preview_img = image[idx:idx + 1]
        img_u8 = (preview_img[0].clamp(0, 1).cpu().numpy() * 255.0 + 0.5).astype(np.uint8)
        ui = {
            "input_image": [_b64_jpeg(img_u8)],
            "w": [int(image.shape[2])],
            "h": [int(image.shape[1])],
            "preview_frame": [int(idx)],
        }

        # Second, high-precision copy of the same frame for the viewer's
        # Inspect mode. The JPEG above is clamped, 8-bit and lossy, so a
        # 16-bit and an 8-bit source are indistinguishable through it; this
        # tile keeps the source's real quantisation steps so a view exposure
        # can reveal banding. See bat_hdr_preview.py. None on failure, and
        # the JS falls back to the JPEG.
        tile = hdr_tile(preview_img[0])
        if tile is not None:
            ui["hdr_tile"] = [tile]
        if mask is not None:
            # Same-frame index for the mask, again clamped to its own
            # batch length (mask may be a single-frame mask used across
            # the whole batch).
            midx = max(0, min(int(preview_frame), mask.shape[0] - 1))
            preview_mask = mask[midx:midx + 1]
            mask_u8 = (preview_mask[0].clamp(0, 1).cpu().numpy() * 255.0 + 0.5).astype(np.uint8)
            ui["input_mask"] = [_b64_png_l(mask_u8)]

        # No .cpu(): forcing the result to host memory cost a full device->host
        # copy (~750MB for a 300-frame 1080p batch) that the next node has to
        # copy straight back. ComfyUI handles device placement.
        return {"ui": ui, "result": (out,)}

File: web/bat_grade.js
/**
 * Bat_Grade — live canvas preview of the grade settings.
 *
 * Workflow:
 *   1. User runs the node once → backend pushes a small JPEG of the
 *      ungraded input (and optional mask) to the canvas.
 *   2. As the artist drags the blackpoint/whitepoint/lift/gain/multiply/
 *      offset/gamma sliders, the JS applies the same Nuke-style formula
 *      to that thumbnail pixel-by-pixel and redraws.
 *   3. The actual graded full-resolution output is computed by the
 *      Python on the NEXT run — the preview is a fast approximation,
 *      not the bit-exact result.
 *
 * Why pixel-by-pixel JS instead of WebGL? A 384px thumbnail is ~150 k
 * pixels; the formula is ~10 multiply-adds per channel; modern V8
 * crunches that in well under 16 ms which is plenty for slider drags.
 * WebGL would be faster but pulls in a lot of boilerplate for no
 * perceptible win at this size.
 */

import { app } from "../../scripts/app.js";
import { addBatDOMWidget, clampNodeSize } from "./bat_node_layout.js";
import {
    hdrSupported, decodeHdrTile, imageDataToSource, buildInspectBar,
} from "./bat_hdr_preview.js";

const NODE_TYPE = "Bat_Grade";

// localStorage preview cache, keyed by node id. Mirrors the pattern in
// bat_roto.js / bat_animated_crop.js — gives the live canvas something
// to show on workflow reopen without bloating the workflow JSON.
function _previewCacheKey(node) {
    return `bat_grade_preview_${node?.id ?? "_"}`;
}
function _saveCachedPreview(node, data) {
    try { localStorage.setItem(_previewCacheKey(node), JSON.stringify(data)); }
    catch (_) {}
}
function _loadCachedPreview(node) {
    try {
        const raw = localStorage.getItem(_previewCacheKey(node));
        return raw ? JSON.parse(raw) : null;
    } catch (_) { return null; }
}

function buildPreview(node) {
    const root = document.createElement("div");
    root.style.cssText = `
        position:relative; display:flex; flex-direction:column;
        background:#0a0a0a; border:1px solid #2a2a2a; border-radius:4px;
        overflow:hidden;
    `;

    const canvas = document.createElement("canvas");
    canvas.style.cssText = "width:100%; height:auto; display:block; background:#000;";
    root.appendChild(canvas);

    const hint = document.createElement("div");
    hint.style.cssText = `
        position:absolute; left:6px; bottom:4px; font:11px monospace;
        color:#9aa; pointer-events:none; text-shadow:0 1px 2px #000;
    `;
    hint.textContent = "Run once to populate the preview.";
    root.appendChild(hint);

    const ctx = canvas.getContext("2d", { willReadFrequently: true });

    // Cached sources so we don't re-decode on every slider drag. Both are the
    // SourceBuffer shape from bat_hdr_preview.js — RGB interleaved Float32,
    // stride 3 — so the pixel loop below never branches on which one is live.
    const state = {
        sourceBitmap: null,    // <img> of the ungraded input (kept for debugging)
        source8:      null,    // SourceBuffer decoded from the 8-bit JPEG
        sourceHdr:    null,    // SourceBuffer decoded from the 16-bit tile, or null
        maskImg:      null,    // <img> of the input mask, or null
        maskCache:    new Map(),   // "WxH" -> Float32Array, rasterised on demand
    };
    node._batGradeState = state;

    // The high-precision tile and the JPEG are different sizes (256px vs 384px
    // long edge), so the mask has to be rasterised per source size rather than
    // once at ingest. Cached because a slider drag repaints every frame.
    function maskFor(w, h) {
        if (!state.maskImg) return null;
        const key = `${w}x${h}`;
        const hit = state.maskCache.get(key);
        if (hit) return hit;
        const c = document.createElement("canvas");
        c.width = w; c.height = h;
        const mctx = c.getContext("2d", { willReadFrequently: true });
        // Stretch to match, same as the backend does before applying.
        mctx.drawImage(state.maskImg, 0, 0, w, h);
        const px = mctx.getImageData(0, 0, w, h).data;
        const out = new Float32Array(w * h);
        for (let i = 0, p = 0; i < out.length; i++, p += 4) out[i] = px[p] / 255;
        state.maskCache.set(key, out);
        return out;
    }

    const inspectBar = buildInspectBar({
        node,
        storageKey: `bat_grade_view_${node?.id ?? "_"}`,
        hasHdr: () => !!state.sourceHdr,
        onChange: () => schedule(),
    });

    function activeSource() {
        return (inspectBar.inspect && state.sourceHdr) ? state.sourceHdr : state.source8;
    }

    // ── widget access ────────────────────────────────────────────────
    const W = (n) => node.widgets.find(w => w.name === n);
    const get = (n) => {
        const w = W(n);
        if (!w) return undefined;
        // BOOLEAN widgets store the value as bool; numeric ones as number.
        return w.value;
    };

    // ── the grade formula — must mirror bat_grade.py's _apply_grade ──
    function applyGrade() {
        const source = activeSource();
        if (!source) return;

        const bp = +get("blackpoint");
        const wp = +get("whitepoint");
        const lift = +get("lift");
        const gain = +get("gain");
        const mult = +get("multiply");
        const off  = +get("offset");
        const gamma = Math.max(+get("gamma"), 0.01);
        const clampW = !!get("clamp_white");
        const clampB = !!get("clamp_black");
        const wpMbp = Math.max(wp - bp, 1e-6);
        const invG  = 1 / gamma;

        // Float32, RGB interleaved (stride 3) whether it came from the JPEG or
        // the 16-bit tile. Grading in float and quantising only at the write
        // below is the whole point: an 8-bit intermediate here would collapse
        // a 16-bit source to the same 256 levels as an 8-bit one, which is
        // exactly what the old Uint8ClampedArray path did.
        const src  = source.data;
        const w    = source.width;
        const h    = source.height;
        const n    = w * h;
        const out  = ctx.createImageData(w, h);
        const dst  = out.data;
        const mask = maskFor(w, h);
        // Viewer exposure — applied after the grade, display only, never part
        // of the rendered result.
        const view = inspectBar.gain;

        for (let i = 0, p = 0, q = 0; i < n; i++, p += 3, q += 4) {
            const r0 = src[p], g0 = src[p+1], b0 = src[p+2];

            // Grade math — one channel at a time, vectorised inline.
            let r = (r0 - bp) / wpMbp;
            let g = (g0 - bp) / wpMbp;
            let b = (b0 - bp) / wpMbp;

            r = r * (gain - lift) + lift;
            g = g * (gain - lift) + lift;
            b = b * (gain - lift) + lift;

            r = r * mult + off;
            g = g * mult + off;
            b = b * mult + off;

            // pow() on negatives is NaN; clamp pre-gamma to >= 0.
            r = r < 0 ? 0 : Math.pow(r, invG);
            g = g < 0 ? 0 : Math.pow(g, invG);
            b = b < 0 ? 0 : Math.pow(b, invG);

            if (clampW) { if (r > 1) r = 1; if (g > 1) g = 1; if (b > 1) b = 1; }
            if (clampB) { if (r < 0) r = 0; if (g < 0) g = 0; if (b < 0) b = 0; }

            // Mask gate (lerp between original and graded by mask value).
            if (mask) {
                const m = mask[i];
                const inv = 1 - m;
                r = r * m + r0 * inv;
                g = g * m + g0 * inv;
                b = b * m + b0 * inv;
            }

            if (view !== 1) { r *= view; g *= view; b *= view; }

            dst[q]   = r > 1 ? 255 : (r < 0 ? 0 : Math.round(r * 255));
            dst[q+1] = g > 1 ? 255 : (g < 0 ? 0 : Math.round(g * 255));
            dst[q+2] = b > 1 ? 255 : (b < 0 ? 0 : Math.round(b * 255));
            dst[q+3] = 255;
        }
        // Resize canvas to match preview size (only if the dimensions changed).
        if (canvas.width !== w || canvas.height !== h) {
            canvas.width = w;
            canvas.height = h;
        }
        ctx.putImageData(out, 0, 0);
        hint.style.display = "none";
    }

    // Schedule a redraw on the next animation frame (coalesces multiple
    // widget callbacks that fire in the same tick — e.g. typing in a
    // number field).
    let raf = 0;
    function schedule() {
        if (raf) return;
        raf = requestAnimationFrame(() => { raf = 0; applyGrade(); });
    }

    // Hook each numeric / boolean widget so any change repaints the
    // preview. Done in onNodeCreated below after the widgets exist; we
    // expose the hook function on the node so the extension can call it.
    node._batGradeWatchWidgets = () => {
        for (const name of ["blackpoint", "whitepoint", "lift", "gain",
                            "multiply", "offset", "gamma",
                            "clamp_white", "clamp_black"]) {
            const w = W(name);
            if (!w) continue;
            const orig = w.callback;
            w.callback = function (v) {
                try { orig?.call(this, v); } catch (_) {}
                schedule();
            };
        }
    };

    // ── ingest helpers (called from onExecuted) ───────────────────────
    node._batGradeIngest = async (b64Image, b64Mask, hdrTile) => {
        const img = new Image();
        await new Promise((res, rej) => {
            img.onload = res; img.onerror = rej;
            img.src = `data:image/jpeg;base64,${b64Image}`;
        });
        state.sourceBitmap = img;
        // Stamp into a backing canvas so we can read pixels.
        const back = document.createElement("canvas");
        back.width = img.naturalWidth;
        back.height = img.naturalHeight;
        const bctx = back.getContext("2d", { willReadFrequently: true });
        bctx.drawImage(img, 0, 0);
        state.source8 = imageDataToSource(
            bctx.getImageData(0, 0, back.width, back.height));

        // Mask is kept as an <img> and rasterised per source size on demand,
        // because the JPEG and the high-precision tile aren't the same size.
        state.maskCache.clear();
        if (b64Mask) {
            const mimg = new Image();
            await new Promise((res, rej) => {
                mimg.onload = res; mimg.onerror = rej;
                mimg.src = `data:image/png;base64,${b64Mask}`;
            });
            state.maskImg = mimg;
        } else {
            state.maskImg = null;
        }

        // High-precision tile. Optional in every direction: the backend may
        // not have produced one, and an old browser may not be able to inflate
        // it — either way the 8-bit path above still works, so a failure here
        // only disables Inspect rather than breaking the preview.
        state.sourceHdr = null;
        if (hdrTile && hdrSupported()) {
            try {
                state.sourceHdr = await decodeHdrTile(hdrTile);
            } catch (e) {
                console.warn("[Bat_Grade] high-precision tile failed to decode:", e);
            }
        }
        inspectBar.refresh();

        // Cache for next reopen — the JPEG only. The 16-bit tile is a few
        // hundred KB and localStorage is a ~5MB origin-wide budget shared with
        // every other BAT node's thumbnail cache, so stashing it there would
        // evict the caches that actually need to survive a reload. Inspect
        // simply waits for the next run.
        _saveCachedPreview(node, { image: b64Image, mask: b64Mask || null });

        schedule();
    };

    // Restore the cached source thumbnail (and mask) on init so the
    // live preview shows something the first time the workflow opens,
    // before any Run.
    node._batGradeRestoreFromCache = () => {
        const cached = _loadCachedPreview(node);
        if (!cached?.image) return;
        node._batGradeIngest(cached.image, cached.mask || null, null);
    };

    root.appendChild(inspectBar.el);
    return root;
}

app.registerExtension({
    name: "Bat_Grade",
    async beforeRegisterNodeDef(nodeType, nodeData, _app) {
        if (nodeData.name !== NODE_TYPE) return;

        const onNodeCreated = nodeType.prototype.onNodeCreated;
        nodeType.prototype.onNodeCreated = function () {
            const r = onNodeCreated ? onNodeCreated.apply(this, arguments) : undefined;
            const el = buildPreview(this);
            // Dual-mode sizing: Nodes 2.0 derives node height from
            // computeLayoutSize, so a bare addDOMWidget + this.size left the
            // node and the widget disagreeing (grey band, no resize).
            addBatDOMWidget(this, "bat_grade_preview", "bat_grade_preview", el, {
                minWidth: 360, height: 520, growable: true,
            });
            this._batGradeWatchWidgets?.();
            clampNodeSize(this, 360, 520);
            // Defer the cache restore until after node.id is finalized
            // (LiteGraph sets it as part of construction; using a 0-ms
            // setTimeout puts us at the back of the microtask queue).
            setTimeout(() => this._batGradeRestoreFromCache?.(), 0);
            return r;
        };

        const onExecuted = nodeType.prototype.onExecuted;
        nodeType.prototype.onExecuted = function (message) {
            const r = onExecuted ? onExecuted.apply(this, arguments) : undefined;
            if (!message || !this._batGradeIngest) return r;
            const one = (v) => Array.isArray(v) ? v[0] : v;
            const img = one(message.input_image);
            const msk = one(message.input_mask);
            const hdr = one(message.hdr_tile);
            if (img) this._batGradeIngest(img, msk, hdr);
            return r;
        };
    },
});
